File: userland/apps/textr/editor.c
#include <ctype.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include "textr.h"

void initEditor(Editor *e) {
    e->mode             = MODE_NORMAL;
    e->cursor.x         = 0;
    e->cursor.y         = 0;
    e->cursor.rx        = 0;
    e->viewport.row_off = 0;
    e->viewport.col_off = 0;
    e->buffer.num_rows  = 0;
    e->buffer.rows      = NULL;
    e->buffer.dirty     = false;
    e->buffer.filename  = NULL;
    e->ui.msg[0]        = '\0';
    e->ui.msg_time      = 0;
    e->ui.cmdline[0]    = '\0';
    e->syntax           = NULL;
    if (getWindowSize(&e->viewport.height, &e->viewport.width) == -1) die("getWindowSize");
    e->viewport.height -= 2; // these rows are reserved for the status bar(s) at the bottom
}

void repositionCursorTL(WriteBuf *wb) {
    writeBufAppend(wb, CURSOR_TL, 3);
}

/* draw fn for entire editor */
void drawRows(WriteBuf *wb) {
    int y;
    for (y = 0; y < E.viewport.height - 1; y++) {
        int filerow = y + E.viewport.row_off;
        if (filerow >= E.buffer.num_rows) {
            if (E.buffer.num_rows == 0 && y == E.viewport.height / 3) {
                /* welcome screen rendered when no file is selected */
                welcomeScreen(wb);
            } else {
                writeBufAppend(wb, "~", 1);
            }
        } else {
            int len = E.buffer.rows[filerow].rsize - E.viewport.col_off;
            if (len < 0) len = 0;
            if (len > E.viewport.width) len = E.viewport.width;
            char          *c          = &E.buffer.rows[filerow].render[E.viewport.col_off];
            unsigned char *hl         = &E.buffer.rows[filerow].hl[E.viewport.col_off];
            char          *curr_color = NULL;
            int            j;
            for (j = 0; j < len; j++) {
                if (iscntrl(c[j])) {
                    char sym = (c[j] <= 26) ? '@' + c[j] : '?';
                    writeBufAppend(wb, "\x1b[7m", 4);
                    writeBufAppend(wb, &sym, 1);
                    writeBufAppend(wb, "\x1b[m", 3);
                    if (curr_color != NULL) { writeBufAppend(wb, curr_color, strlen(curr_color)); }
                } else if (hl[j] == HL_NORMAL) {
                    if (curr_color != NULL) {
                        writeBufAppend(wb, DEF_COLOR, 5);
                        curr_color = NULL;
                    }
                    writeBufAppend(wb, &c[j], 1);
                } else {
                    char *color = editorSyntaxToColor(hl[j]);
                    if (color != curr_color) {
                        curr_color = color;
                        writeBufAppend(wb, color, strlen(color));
                    }
                    writeBufAppend(wb, &c[j], 1);
                }
            }
            writeBufAppend(wb, DEF_COLOR, 5);
        }

        writeBufAppend(wb, "\x1b[K", 3);
        writeBufAppend(wb, "\r\n", 2);
    }
}

void welcomeScreen(WriteBuf *wb) {
    char welcome_title[50];
    char welcome_desc1[45];
    char welcome_desc2[60];

    int titlelen = snprintf(welcome_title, sizeof(welcome_title), "asemics -- 0.1");
    int desclen1 =
        snprintf(welcome_desc1, sizeof(welcome_desc1), "Asemic - mark-making that resembles text");
    int desclen2 = snprintf(welcome_desc2, sizeof(welcome_desc2),
                            " or handwriting but carries no specific literal meaning.");

    if (titlelen > E.viewport.width) titlelen = E.viewport.width;
    if (desclen1 > E.viewport.width) desclen1 = E.viewport.width;
    if (desclen2 > E.viewport.width) desclen2 = E.viewport.width;

    int title_padding = (E.viewport.width - titlelen) / 2;
    int desc_padding1 = (E.viewport.width - desclen1) / 2;
    int desc_padding2 = (E.viewport.width - desclen2) / 2;

    if (title_padding) {
        writeBufAppend(wb, "~", 1);
        title_padding--;
    }

    /* title */
    while (title_padding--)
        writeBufAppend(wb, " ", 1);
    writeBufAppend(wb, welcome_title, titlelen);
    writeBufAppend(wb, "\r\n\n", 3);

    /* desc 1 */
    while (desc_padding1--)
        writeBufAppend(wb, " ", 1);
    writeBufAppend(wb, welcome_desc1, desclen1);
    writeBufAppend(wb, "\r\n", 2);

    /* desc 2 */
    while (desc_padding2--)
        writeBufAppend(wb, " ", 1);
    writeBufAppend(wb, welcome_desc2, desclen2);
}

void updateCursorShape(WriteBuf *wb) {
    switch (E.mode) {
    case MODE_NORMAL: writeBufAppend(wb, CURSOR_BLOCK, 5); break;
    case MODE_COMMAND: writeBufAppend(wb, CURSOR_HIDE, 6); break;
    case MODE_INSERT: writeBufAppend(wb, CURSOR_BAR, 5); break;
    case MODE_VISUAL: writeBufAppend(wb, CURSOR_BLOCK, 5); break;
    default: return;
    }
}

void refreshScreen(void) {
    editorScroll();
    WriteBuf wb = writeBufInit();

    writeBufAppend(&wb, CURSOR_HIDE, 6);
    writeBufAppend(&wb, SCREEN_CLEAR, 4);
    repositionCursorTL(&wb);
    drawRows(&wb);

    char buf[32];
    int  n;

    n = snprintf(buf, sizeof(buf), "\x1b[%d;1H", E.viewport.height + 1);
    writeBufAppend(&wb, buf, n);
    editorDrawStatusBar(&wb);

    n = snprintf(buf, sizeof(buf), "\x1b[%d;1H", E.viewport.height + 2);
    writeBufAppend(&wb, buf, n);
    if (E.ui.cmdline[0] != '\0') {
        editorDrawCmdline(&wb);
    } else {
        editorDrawMsgBar(&wb);
    }

    n = snprintf(buf, sizeof(buf), "\x1b[%d;%dH", (E.cursor.y - E.viewport.row_off) + 1,
                 (E.cursor.rx - E.viewport.col_off) + 1);
    writeBufAppend(&wb, buf, n);

    /* Enabled cursor and render cursor based on EDITOR MODE */
    writeBufAppend(&wb, CURSOR_SHOW, 6);
    updateCursorShape(&wb);
    write(STDOUT_FILENO, wb.data, wb.len);
    writeBufFree(&wb);
}

void clearScreen(void) {
    write(STDOUT_FILENO, SCREEN_CLEAR, 4);
    write(STDOUT_FILENO, CURSOR_TL, 3);
}

WriteBuf writeBufInit() {
    WriteBuf wb;
    wb.data   = NULL;
    wb.len = 0;
    return wb;
}

void writeBufAppend(WriteBuf *wb, const char *s, int len) {
    char *new = realloc(wb->data, wb->len + len);

    if (new == NULL) return;
    memcpy(&new[wb->len], s, len);
    wb->data = new;
    wb->len += len;
}

void writeBufFree(WriteBuf *wb) {
    free(wb->data);
}

void moveCursor(int key) {
    Row *row = (E.cursor.y >= E.buffer.num_rows) ? NULL : &E.buffer.rows[E.cursor.y];
    switch (key) {
    case 'h':
    case ARROW_LEFT:
        if (E.cursor.x != 0) { E.cursor.x--; }
        break;

    case 'j':
    case ARROW_DOWN:
        if (E.cursor.y < E.buffer.num_rows) { E.cursor.y++; }
        break;

    case 'k':
    case ARROW_UP:
        if (E.cursor.y != 0) { E.cursor.y--; }
        break;

    case 'l':
    case ARROW_RIGHT:
        if (row && E.cursor.x < row->size - 1) { E.cursor.x++; }
        break;
    }

    row        = (E.cursor.y >= E.buffer.num_rows) ? NULL : &E.buffer.rows[E.cursor.y];
    int rowlen = row ? row->size : 0;
    if (E.cursor.x > rowlen) { E.cursor.x = rowlen; }
}

int editorCreateFile(char *filename) {
    int fd = open(filename, O_CREAT | O_WRONLY, 0644);
    if (fd == -1) {
        perror(filename);
        return 1;
    }

    close(fd);
    return 0;
}

void editorOpen(char *filename) {
    free(E.buffer.filename);
    E.buffer.filename = strdup(filename);

    editorSetSyntaxHighlight();

    FILE *fp = fopen(filename, "r");
    if (!fp) {
        int ok = editorCreateFile(filename);
        if (ok != 0) { die("open & create"); }
        fp = fopen(filename, "r");
    }

    char   *line    = NULL;
    size_t  linecap = 0;
    ssize_t linelen;
    while ((linelen = getline(&line, &linecap, fp)) != -1) {
        while (linelen > 0 && (line[linelen - 1] == '\n' || line[linelen - 1] == '\r'))
            linelen--;

        editorInsertRow(E.buffer.num_rows, line, linelen);
    }
    free(line);
    fclose(fp);
    E.buffer.dirty = false;
}

void editorScroll() {
    E.cursor.rx = 0;
    if (E.cursor.y < E.buffer.num_rows) { E.cursor.rx = editorRowXtoRx(&E.buffer.rows[E.cursor.y], E.cursor.x); }

    if (E.cursor.y < E.viewport.row_off) { E.viewport.row_off = E.cursor.y; }
    if (E.cursor.y >= E.viewport.row_off + E.viewport.height) { E.viewport.row_off = E.cursor.y - E.viewport.height + 1; }
    if (E.cursor.rx < E.viewport.col_off) { E.viewport.col_off = E.cursor.rx; }
    if (E.cursor.rx >= E.viewport.col_off + E.viewport.width) { E.viewport.col_off = E.cursor.rx - E.viewport.width + 1; }
}

char *editorRowsToStr(int *buflen) {
    int totallen = 0;
    int j;

    for (j = 0; j < E.buffer.num_rows; j++)
        totallen += E.buffer.rows[j].size + 1;
    *buflen = totallen;

    char *buf = malloc(totallen);
    char *p   = buf;

    for (j = 0; j < E.buffer.num_rows; j++) {
        memcpy(p, E.buffer.rows[j].chars, E.buffer.rows[j].size);
        p += E.buffer.rows[j].size;
        *p = '\n';
        p++;
    }

    return buf;
}

void editorSave() {
    if (E.buffer.filename == NULL) return;

    int   len;
    char *buf = editorRowsToStr(&len);

    int fd = open(E.buffer.filename, O_RDWR | O_CREAT, 0644);
    if (fd != -1) {
        if (ftruncate(fd, len) != -1) {
            if (write(fd, buf, len) == len) {
                close(fd);
                free(buf);
                E.buffer.dirty   = false;
                E.ui.cmdline[0]  = '\0';
                editorSetStatusMsg("%d: bytes written to disk", len);
                editorSetSyntaxHighlight();
                return;
            }
            editorSetSyntaxHighlight();
        }
        close(fd);
    }
    free(buf);
    editorSetStatusMsg("Failed to save. I/O error: %s", strerror(errno));
}

void editorQuit(bool force) {
    if (force) {
        clearScreen();
        exit(0);
    } else {
        if (E.buffer.dirty) {
            char buf[80];
            int  n = snprintf(buf, sizeof(buf),
                              "%s has unsaved changes. '!q' to quit without saving", E.buffer.filename);
            int  i = 0;
            E.ui.cmdline[0] = '\0';
            while (buf[i] != '\0') {
                commandInsertChar(buf[i]);
                i++;
            }
        } else {
            clearScreen();
            exit(0);
        }
    }
}

char *editorPrompt(char *prompt, void (*callback)(char *, int)) {
    size_t bufsize = 80;
    char  *buf     = malloc(bufsize);

    size_t buflen = 0;
    buf[0]        = '\0';

    while (1) {
        editorSetStatusMsg(prompt, buf);
        refreshScreen();

        int c = readKey();
        if (c == BACKSPACE) {
            if (buflen != 0) buf[--buflen] = '\0';
        } else if (c == '\x1b') {
            editorSetStatusMsg("");
            if (callback) callback(buf, c);
            free(buf);
            return NULL;
        } else if (c == '\r') {
            if (buflen != 0) {
                editorSetStatusMsg("");
                if (callback) callback(buf, c);
                return buf;
            }
        } else if (!iscntrl(c) && c < 128) {
            if (buflen == bufsize - 1) {
                bufsize *= 2;
                buf = realloc(buf, bufsize);
            }
            buf[buflen++] = c;
            buf[buflen]   = '\0';
        }

        if (callback) callback(buf, c);
    }
}

File: userland/apps/asemics-rs/src/editor.rs
use std::fs;
use std::io::Write;
use std::os::unix::fs::OpenOptionsExt;
use std::time::SystemTime;

use crate::consts::*;
use crate::dbg::init_dbg;
use crate::history::History;
use crate::input::{read_key, BACKSPACE, ARROW_DOWN, ARROW_LEFT, ARROW_RIGHT, ARROW_UP};
use crate::rows::Row;
use crate::syntax::{syntax_to_color, EditorSyntax, HL_NORMAL};
use crate::terminal::{die, exit_restore, get_window_size, strerror_last, write_stdout};

/* global state */
#[derive(Clone, Copy, PartialEq)]
pub enum EditorMode {
    Normal,
    Insert,
    Command,
    Visual,
}

#[derive(Clone, Copy)]
pub struct Cursor {
    pub x: i32,  /* cursor column in the buffer (logical char index, ignores tabs) */
    pub y: i32,  /* cursor row in the buffer (line index) */
    pub rx: i32, /* cursor column on screen (`x` with tabs expanded) */
}

pub struct Viewport {
    pub height: i32,  /* number of visible rows on screen */
    pub width: i32,   /* number of visible columns on screen */
    pub row_off: i32, /* topmost visible buffer row (vertical scroll offset) */
    pub col_off: i32, /* leftmost visible column (horizontal scroll offset) */
}

pub struct Buffer {
    pub rows: Vec<Row>, /* the document contents */
    /* kept separately from rows.len(): the C code bumps num_rows only after the
     * syntax update runs during a row insert, and comment propagation reads it */
    pub num_rows: i32,
    pub dirty: bool,              /* true if there are unsaved changes */
    pub filename: Option<String>, /* path to the open file (None when unnamed) */
}

pub struct StatusBar {
    pub msg: Vec<u8>,            /* transient status message shown at the bottom */
    pub msg_time: SystemTime,    /* when `msg` was set, used to time it out */
    pub cmdline: Vec<u8>,        /* text being typed in command mode (`:w`, `:q`, ...) */
}

/* position struct for motions */
#[derive(Clone, Copy)]
pub struct Pos {
    pub x: i32,
    pub y: i32,
}

pub struct Editor {
    pub mode: EditorMode,                      /* current modal state */
    pub cursor: Cursor,                        /* where the cursor is */
    pub viewport: Viewport,                    /* what slice of the buffer is visible */
    pub buffer: Buffer,                        /* the document being edited */
    pub ui: StatusBar,                         /* bottom-bar state: status message + command line */
    pub syntax: Option<&'static EditorSyntax>, /* active syntax-highlight rules (None = none) */
    pub history: History,
    pub dbg: fs::File,
}

pub fn is_cntrl_byte(c: u8) -> bool {
    c < 32 || c == 127
}

/* iscntrl(3) as the C code uses it: readKey returns a sign-extended char, so
 * bytes >= 128 arrive negative and are not control chars */
pub fn is_cntrl_int(c: i32) -> bool {
    (0..32).contains(&c) || c == 127
}

impl Editor {
    pub fn new() -> Editor {
        let dbg = init_dbg();

        let Some((mut height, mut width)) = get_window_size() else {
            die("getWindowSize")
        };
        height -= STATUS_BAR_RESERVE; // status bar space for bottom of screen
        width -= LINE_NUM_RESERVE;

        Editor {
            mode: EditorMode::Normal,
            cursor: Cursor { x: 0, y: 0, rx: 0 },
            viewport: Viewport {
                height,
                width,
                row_off: 0,
                col_off: 0,
            },
            buffer: Buffer {
                rows: Vec::new(),
                num_rows: 0,
                dirty: false,
                filename: None,
            },
            ui: StatusBar {
                msg: Vec::new(),
                msg_time: SystemTime::UNIX_EPOCH,
                cmdline: Vec::new(),
            },
            syntax: None,
            history: History::new(),
            dbg,
        }
    }

    /* draw fn for entire editor */
    pub fn draw_rows(&self, wb: &mut Vec<u8>) {
        for y in 0..self.viewport.height {
            let filerow = y + self.viewport.row_off;

            if filerow >= self.buffer.num_rows {
                if self.buffer.num_rows == 0 && y == self.viewport.height / 3 {
                    /* welcome screen rendered when no file is selected */
                    self.welcome_screen(wb);
                } else {
                    wb.push(b'~');
                }
            } else {
                self.draw_line_nums(wb, filerow);

                let row = &self.buffer.rows[filerow as usize];
                let mut len = row.rsize() - self.viewport.col_off;
                if len < 0 {
                    len = 0;
                }
                if len > self.viewport.width {
                    len = self.viewport.width;
                }

                let start = self.viewport.col_off as usize;
                let mut curr_color: Option<&'static str> = None;

                for j in 0..len as usize {
                    let c = row.render[start + j];
                    let hl = row.hl[start + j];

                    if is_cntrl_byte(c) {
                        let sym = if c <= 26 { b'@' + c } else { b'?' };
                        wb.extend_from_slice(b"\x1b[7m");
                        wb.push(sym);
                        wb.extend_from_slice(b"\x1b[m");

                        if let Some(color) = curr_color {
                            wb.extend_from_slice(color.as_bytes());
                        }
                    } else if hl == HL_NORMAL {
                        if curr_color.is_some() {
                            wb.extend_from_slice(DEF_COLOR.as_bytes());
                            curr_color = None;
                        }
                        wb.push(c);
                    } else {
                        let color = syntax_to_color(hl);
                        if Some(color) != curr_color {
                            curr_color = Some(color);
                            wb.extend_from_slice(color.as_bytes());
                        }
                        wb.push(c);
                    }
                }
                wb.extend_from_slice(DEF_COLOR.as_bytes());
            }

            wb.extend_from_slice(SCREEN_CLEAR_LINE);
            wb.extend_from_slice(b"\r\n");
        }
    }

    pub fn draw_line_nums(&self, wb: &mut Vec<u8>, filerow: i32) {
        /* highlight current line num for cursor */
        let gutter = if self.cursor.y == filerow {
            format!("{:5}  ", filerow + 1)
        } else {
            format!("{}{:5}  {}", DARK_GRAY, filerow + 1, RESET_FG)
        };
        wb.extend_from_slice(gutter.as_bytes());
    }

    pub fn welcome_screen(&self, wb: &mut Vec<u8>) {
        let welcome_title = "asemics -- 0.1";
        let welcome_desc1 = "Asemic - mark-making that resembles text";
        let welcome_desc2 = " or handwriting but carries no specific literal meaning.";

        let mut titlelen = welcome_title.len() as i32;
        let mut desclen1 = welcome_desc1.len() as i32;
        let mut desclen2 = welcome_desc2.len() as i32;

        if titlelen > self.viewport.width {
            titlelen = self.viewport.width;
        }
        if desclen1 > self.viewport.width {
            desclen1 = self.viewport.width;
        }
        if desclen2 > self.viewport.width {
            desclen2 = self.viewport.width;
        }

        let mut title_padding = (self.viewport.width - titlelen) / 2;
        let desc_padding1 = (self.viewport.width - desclen1) / 2;
        let desc_padding2 = (self.viewport.width - desclen2) / 2;

        wb.extend_from_slice(CURSOR_HIDE);

        if title_padding != 0 {
            wb.push(b'~');
            title_padding -= 1;
        }

        /* title */
        for _ in 0..title_padding {
            wb.push(b' ');
        }
        wb.extend_from_slice(&welcome_title.as_bytes()[..titlelen as usize]);
        wb.extend_from_slice(b"\r\n\n");

        /* desc 1 */
        for _ in 0..desc_padding1 {
            wb.push(b' ');
        }
        wb.extend_from_slice(&welcome_desc1.as_bytes()[..desclen1 as usize]);
        wb.extend_from_slice(b"\r\n");

        /* desc 2 */
        for _ in 0..desc_padding2 {
            wb.push(b' ');
        }
        wb.extend_from_slice(&welcome_desc2.as_bytes()[..desclen2 as usize]);
    }

    pub fn update_cursor_shape(&self, wb: &mut Vec<u8>) {
        match self.mode {
            EditorMode::Normal => wb.extend_from_slice(CURSOR_BLOCK),
            EditorMode::Command => wb.extend_from_slice(CURSOR_HIDE),
            EditorMode::Insert => wb.extend_from_slice(CURSOR_BAR),
            EditorMode::Visual => wb.extend_from_slice(CURSOR_BLOCK),
        }
    }

    pub fn refresh_screen(&mut self) {
        self.scroll();
        let mut wb: Vec<u8> = Vec::new();

        wb.extend_from_slice(CURSOR_HIDE);
        wb.extend_from_slice(SCREEN_CLEAR);
        wb.extend_from_slice(CURSOR_TL);
        self.draw_rows(&mut wb);

        /* status bar buffer */
        wb.extend_from_slice(format!("\x1b[{};1H", self.viewport.height + 1).as_bytes());
        self.draw_status_bar(&mut wb);

        /* message bar buffer */
        wb.extend_from_slice(format!("\x1b[{};1H", self.viewport.height + 2).as_bytes());
        if !self.ui.cmdline.is_empty() {
            self.draw_cmdline(&mut wb);
        } else {
            self.draw_msg_bar(&mut wb);
        }

        /* line number gutter buffer */
        wb.extend_from_slice(
            format!(
                "\x1b[{};{}H",
                (self.cursor.y - self.viewport.row_off) + 1,
                (self.cursor.rx - self.viewport.col_off) + LINE_NUM_RESERVE // consumes 8 visual cols
            )
            .as_bytes(),
        );

        /* Enabled cursor and render cursor based on EDITOR MODE && if a file is open */
        if self.buffer.filename.is_some() {
            wb.extend_from_slice(CURSOR_SHOW);
            self.update_cursor_shape(&mut wb);
        } else {
            wb.extend_from_slice(CURSOR_HIDE);
        }

        write_stdout(&wb);
    }

    pub fn move_cursor(&mut self, key: i32) {
        let row_size = |e: &Editor| -> Option<i32> {
            if e.cursor.y >= e.buffer.num_rows {
                None
            } else {
                Some(e.buffer.rows[e.cursor.y as usize].size())
            }
        };

        let row = row_size(self);
        if key == b'h' as i32 || key == ARROW_LEFT {
            if self.cursor.x != 0 {
                self.cursor.x -= 1;
            }
        } else if key == b'j' as i32 || key == ARROW_DOWN {
            if self.cursor.y < self.buffer.num_rows - 1 {
                self.cursor.y += 1;
            }
        } else if key == b'k' as i32 || key == ARROW_UP {
            if self.cursor.y != 0 {
                self.cursor.y -= 1;
            }
        } else if key == b'l' as i32 || key == ARROW_RIGHT {
            if let Some(size) = row {
                if self.cursor.x < size - 1 {
                    self.cursor.x += 1;
                }
            }
        }

        let rowlen = row_size(self).unwrap_or(0);
        if self.cursor.x > rowlen {
            self.cursor.x = rowlen;
        }
    }

    pub fn create_file(filename: &str) -> i32 {
        match fs::OpenOptions::new()
            .create(true)
            .write(true)
            .mode(0o644)
            .open(filename)
        {
            Ok(_) => 0,
            Err(_) => {
                eprintln!("{}: {}", filename, strerror_last());
                1
            }
        }
    }

    pub fn open(&mut self, filename: &str) {
        self.buffer.filename = Some(filename.to_string());

        self.set_syntax_highlight();

        let data = match fs::read(filename) {
            Ok(d) => d,
            Err(_) => {
                if Editor::create_file(filename) != 0 {
                    die("open & create");
                }
                match fs::read(filename) {
                    Ok(d) => d,
                    Err(_) => die("open & create"),
                }
            }
        };

        for chunk in data.split_inclusive(|&b| b == b'\n') {
            let mut line = chunk;
            while let Some((&last, rest)) = line.split_last() {
                if last == b'\n' || last == b'\r' {
                    line = rest;
                } else {
                    break;
                }
            }
            self.insert_row(self.buffer.num_rows, line);
        }
        self.buffer.dirty = false;
    }

    /* y = 0: top of file  */
    pub fn scroll(&mut self) {
        /* obtain cursor column */
        self.cursor.rx = 0;
        if self.cursor.y < self.buffer.num_rows {
            self.cursor.rx = self.buffer.rows[self.cursor.y as usize].x_to_rx(self.cursor.x);
        }

        /* scrolloff for approaching top of file */
        if self.cursor.y < self.viewport.row_off + SCROLL_OFF {
            self.viewport.row_off = self.cursor.y - SCROLL_OFF;
            if self.viewport.row_off < 0 {
                self.viewport.row_off = 0;
            }
        }

        /* scrolloff for approaching bottom of file */
        if self.cursor.y >= self.viewport.row_off + self.viewport.height - SCROLL_OFF {
            self.viewport.row_off = self.cursor.y - self.viewport.height + 1 + SCROLL_OFF;
        }

        if self.cursor.rx < self.viewport.col_off {
            self.viewport.col_off = self.cursor.rx;
        }
        if self.cursor.rx >= self.viewport.col_off + self.viewport.width {
            self.viewport.col_off = self.cursor.rx - self.viewport.width + 1;
        }
    }

    pub fn rows_to_string(&self) -> Vec<u8> {
        let mut buf = Vec::new();
        for row in &self.buffer.rows[..self.buffer.num_rows as usize] {
            buf.extend_from_slice(&row.chars);
            buf.push(b'\n');
        }
        buf
    }

    pub fn save(&mut self) {
        let Some(filename) = self.buffer.filename.clone() else {
            return;
        };

        let buf = self.rows_to_string();
        let len = buf.len();

        if let Ok(mut file) = fs::OpenOptions::new()
            .read(true)
            .write(true)
            .create(true)
            .mode(0o644)
            .open(&filename)
        {
            if file.set_len(len as u64).is_ok() {
                if file.write_all(&buf).is_ok() {
                    self.buffer.dirty = false;
                    self.ui.cmdline.clear();
                    self.set_status_msg(format!("{}: bytes written to disk", len));
                    self.set_syntax_highlight();
                    return;
                }
                self.set_syntax_highlight();
            }
        }
        self.set_status_msg(format!("Failed to save. I/O error: {}", strerror_last()));
    }

    pub fn quit(&mut self, force: bool) {
        if force {
            clear_screen();
            exit_restore(0);
        } else if self.buffer.dirty {
            let msg = format!(
                "{} has unsaved changes. '!q' to quit without saving",
                self.buffer.filename.as_deref().unwrap_or("(null)")
            );
            let mut msg = msg.into_bytes();
            msg.truncate(79); // C builds this in a char buf[80]
            self.ui.cmdline.clear();
            for &b in &msg {
                self.command_insert_char(b as i32);
            }
        } else {
            clear_screen();
            exit_restore(0);
        }
    }

    pub fn prompt(
        &mut self,
        prompt_fmt: &str,
        mut callback: impl FnMut(&mut Editor, &[u8], i32),
    ) -> Option<Vec<u8>> {
        let mut buf: Vec<u8> = Vec::new();

        loop {
            let shown = prompt_fmt.replace("%s", &String::from_utf8_lossy(&buf));
            self.set_status_msg(shown);
            self.refresh_screen();

            let c = read_key();
            if c == BACKSPACE {
                if !buf.is_empty() {
                    buf.pop();
                }
            } else if c == ESC {
                self.set_status_msg(String::new());
                callback(self, &buf, c);
                return None;
            } else if c == ENTER {
                if !buf.is_empty() {
                    self.set_status_msg(String::new());
                    callback(self, &buf, c);
                    return Some(buf);
                }
            } else if !is_cntrl_int(c) && c < 128 {
                buf.push(c as u8);
            }

            callback(self, &buf, c);
        }
    }
}

pub fn clear_screen() {
    write_stdout(SCREEN_CLEAR);
    write_stdout(CURSOR_TL);
}
